File: sdk/python/easynet_sdk/providers/easynet/lifecycle.py
"""EasyNet provider lifecycle policy and source-compatible daemon DTOs."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from enum import StrEnum
from typing import TYPE_CHECKING, Mapping

from ...connection import ConnectOptions
from ...errors import ErrorCode, RetryHint, SDKError

if TYPE_CHECKING:
    from ...runtime_lifecycle import (
        AttachOptions,
        Endpoints,
        RuntimeHandle,
        RuntimeLifecycleTransport,
    )
    from ...runtime import RuntimeClient


class DaemonMode(StrEnum):
    """easynet-daemon deployment role."""

    DEVICE = "device"
    HUB = "hub"
    BOTH = "both"


@dataclass(frozen=True)
class StartConfig:
    """easynet-daemon process start policy."""

    mode: DaemonMode
    realm: str = ""
    device_id: str = ""
    home_dir: str = ""
    daemon_bin: str = ""
    log_path: str = ""
    detached: bool = False
    env: Mapping[str, str] = field(default_factory=dict)
    uds_path: str = ""
    listen_tcp: str = ""
    tls_cert_path: str = ""
    tls_key_path: str = ""
    hub_endpoint: str = ""
    trust_path: str = ""

    def validate(self) -> None:
        if self.mode == DaemonMode.DEVICE and self.listen_tcp.strip():
            raise _invalid_lifecycle(
                "device mode must not accept a public TCP listener"
            )
        if (
            self.mode in {DaemonMode.HUB, DaemonMode.BOTH}
            and self.listen_tcp.strip()
            and (not self.tls_cert_path.strip() or not self.tls_key_path.strip())
        ):
            raise _invalid_lifecycle("public TCP listener requires TLS material")

    def to_json_dict(self) -> dict[str, object]:
        value: dict[str, object] = {"mode": self.mode.value}
        for key in (
            "realm",
            "device_id",
            "home_dir",
            "daemon_bin",
            "log_path",
            "uds_path",
            "listen_tcp",
            "tls_cert_path",
            "tls_key_path",
            "hub_endpoint",
            "trust_path",
        ):
            item = getattr(self, key)
            if item:
                value[key] = item
        if self.detached:
            value["detached"] = True
        if self.env:
            value["env"] = dict(self.env)
        return value

    def to_json_bytes(self) -> bytes:
        return json.dumps(
            self.to_json_dict(), separators=(",", ":"), sort_keys=True
        ).encode("utf-8")


@dataclass(frozen=True)
class DiscoverOptions:
    """easynet-daemon discovery policy, including its product HOME root."""

    control_endpoint: str = ""
    control_path: str = ""
    home_dir: str = ""

    def to_json_bytes(self) -> bytes:
        return _json_bytes(
            {
                "control_endpoint": self.control_endpoint,
                "control_path": self.control_path,
                "home_dir": self.home_dir,
            }
        )


def start_daemon(
    transport: "RuntimeLifecycleTransport", config: StartConfig
) -> "RuntimeHandle":
    from ...runtime_lifecycle import start_runtime_host

    return start_runtime_host(transport, config)


def attach_daemon(
    transport: "RuntimeLifecycleTransport",
    options: "AttachOptions | None" = None,
) -> "RuntimeHandle":
    from ...runtime_lifecycle import AttachOptions, attach_runtime_host

    return attach_runtime_host(transport, options or AttachOptions())


def discover_daemon(
    transport: "RuntimeLifecycleTransport",
    options: DiscoverOptions = DiscoverOptions(),
) -> "Endpoints":
    from ...runtime_lifecycle import discover_runtime_host

    return discover_runtime_host(transport, options)


def connect_local(
    transport: "RuntimeLifecycleTransport",
    options: ConnectOptions = ConnectOptions(),
) -> "RuntimeClient":
    from ...runtime_lifecycle import connect_runtime_local

    return connect_runtime_local(transport, options)


def _invalid_lifecycle(message: str) -> SDKError:
    return SDKError(
        code=ErrorCode.INVALID_ARGUMENT,
        stage="sdk",
        retry=RetryHint.NEVER,
        retryable=False,
        message=message,
    )


def _json_bytes(value: Mapping[str, object]) -> bytes:
    compact = {key: item for key, item in value.items() if item not in ("", 0, False)}
    return json.dumps(compact, separators=(",", ":"), sort_keys=True).encode("utf-8")

File: sdk/go/provider/easynet/contract/lifecycle.go
// Package contract owns the EasyNet provider's daemon lifecycle wire contract.
//
// Runtime lifecycle state and handle ownership remain in the canonical SDK.
package contract

import (
	"encoding/json"
	"errors"
	"strings"
)

// Mode is the easynet-daemon deployment role.
type Mode string

const (
	ModeDevice Mode = "device"
	ModeHub    Mode = "hub"
	ModeBoth   Mode = "both"
)

// StartConfig is the easynet-daemon process start policy.
type StartConfig struct {
	Mode        Mode              `json:"mode"`
	Realm       string            `json:"realm,omitempty"`
	DeviceID    string            `json:"device_id,omitempty"`
	HomeDir     string            `json:"home_dir,omitempty"`
	DaemonBin   string            `json:"daemon_bin,omitempty"`
	WorkingDir  string            `json:"working_dir,omitempty"`
	LogPath     string            `json:"log_path,omitempty"`
	Detached    bool              `json:"detached,omitempty"`
	Env         map[string]string `json:"env,omitempty"`
	UDSPath     string            `json:"uds_path,omitempty"`
	ListenTCP   string            `json:"listen_tcp,omitempty"`
	TLSCertPath string            `json:"tls_cert_path,omitempty"`
	TLSKeyPath  string            `json:"tls_key_path,omitempty"`
	HubEndpoint string            `json:"hub_endpoint,omitempty"`
	TrustPath   string            `json:"trust_path,omitempty"`
}

// AttachOptions identifies an existing easynet-daemon process.
type AttachOptions struct {
	ControlEndpoint    string `json:"control_endpoint,omitempty"`
	InvocationEndpoint string `json:"invocation_endpoint,omitempty"`
	ControlPath        string `json:"control_path,omitempty"`
}

// DiscoverOptions controls easynet-daemon discovery.
type DiscoverOptions struct {
	ControlEndpoint string `json:"control_endpoint,omitempty"`
	ControlPath     string `json:"control_path,omitempty"`
	HomeDir         string `json:"home_dir,omitempty"`
}

// StopOptions controls easynet-daemon shutdown.
type StopOptions struct {
	GracefulTimeoutMS int64 `json:"graceful_timeout_ms,omitempty"`
	Force             bool  `json:"force,omitempty"`
}

// ValidationError reports invalid EasyNet provider policy before transport use.
type ValidationError struct {
	message string
}

func (e *ValidationError) Error() string {
	if e == nil {
		return ""
	}
	return e.message
}

// Validate enforces easynet-daemon deployment policy.
func (c StartConfig) Validate() error {
	switch c.Mode {
	case ModeDevice:
		if strings.TrimSpace(c.ListenTCP) != "" {
			return &ValidationError{message: "device mode must not accept a public TCP listener"}
		}
	case ModeHub, ModeBoth:
		if strings.TrimSpace(c.ListenTCP) != "" &&
			(strings.TrimSpace(c.TLSCertPath) == "" || strings.TrimSpace(c.TLSKeyPath) == "") {
			return &ValidationError{message: "public TCP listener requires TLS material"}
		}
	default:
		return &ValidationError{message: "mode must be device, hub, or both"}
	}
	return nil
}

// RuntimeHostStartPayload lowers validated EasyNet process policy for the
// canonical runtime-host lifecycle transport.
func (c StartConfig) RuntimeHostStartPayload() ([]byte, error) {
	return json.Marshal(c)
}

// RuntimeHostDiscoverPayload lowers EasyNet directory policy for canonical
// endpoint discovery.
func (o DiscoverOptions) RuntimeHostDiscoverPayload() ([]byte, error) {
	return json.Marshal(o)
}

// IsValidationError reports whether an error came from EasyNet provider policy.
func IsValidationError(err error) bool {
	var target *ValidationError
	return errors.As(err, &target)
}
